pkg/rtc/transportmanager.go

```diff
@@ -0,0 +1,514 @@
+	pcTransport := t.publisher
+	if (isPrimary && t.params.SubscriberAsPrimary) || (!isPrimary && !t.params.SubscriberAsPrimary) {
+		pcTransport = t.subscriber
+	}
+
+	return pcTransport
+}
+
+func (t *TransportManager) handleConnectionFailed(isShortLived bool) {
+	if !t.params.AllowTCPFallback || !isShortLived {
+		return
+	}
+
+	// irrespective of which one fails, force TCP on both as the other one might
+	// fail at a different time and cause another disruption
+	t.params.Logger.Infow("restricting transport to TCP on both peer connections")
+	t.SetICEConfig(types.IceConfig{
+		PreferPubTcp: true,
+		PreferSubTcp: true,
+	})
+}
+
+func (t *TransportManager) SetMigrateInfo(previousAnswer *webrtc.SessionDescription, dataChannels []*livekit.DataChannelInfo) {
+	t.lock.Lock()
+	t.pendingDataChannelsPublisher = make([]*livekit.DataChannelInfo, 0, len(dataChannels))
+	pendingDataChannelsSubscriber := make([]*livekit.DataChannelInfo, 0, len(dataChannels))
+	for _, dci := range dataChannels {
+		if dci.Target == livekit.SignalTarget_SUBSCRIBER {
+			pendingDataChannelsSubscriber = append(pendingDataChannelsSubscriber, dci)
+		} else {
+			t.pendingDataChannelsPublisher = append(t.pendingDataChannelsPublisher, dci)
+		}
+	}
+	t.lock.Unlock()
+
+	if t.params.SubscriberAsPrimary {
+		if err := t.createDataChannelsForSubscriber(pendingDataChannelsSubscriber); err != nil {
+			t.params.Logger.Errorw("create subscriber data channels during migration failed", err)
+		}
+	}
+
+	t.subscriber.SetPreviousAnswer(previousAnswer)
+}
+
+func (t *TransportManager) ProcessPendingPublisherDataChannels() {
+	t.lock.Lock()
+	pendingDataChannels := t.pendingDataChannelsPublisher
+	t.pendingDataChannelsPublisher = nil
+	t.lock.Unlock()
+
+	ordered := true
+	negotiated := true
+
+	for _, ci := range pendingDataChannels {
+		var (
+			dcLabel    string
+			dcID       uint16
+			dcExisting bool
+			err        error
+		)
+		if ci.Label == LossyDataChannel {
+			retransmits := uint16(0)
+			id := uint16(ci.GetId())
+			dcLabel, dcID, dcExisting, err = t.publisher.CreateDataChannelIfEmpty(LossyDataChannel, &webrtc.DataChannelInit{
+				Ordered:        &ordered,
+				MaxRetransmits: &retransmits,
+				Negotiated:     &negotiated,
+				ID:             &id,
+			})
+		} else if ci.Label == ReliableDataChannel {
+			id := uint16(ci.GetId())
+			dcLabel, dcID, dcExisting, err = t.publisher.CreateDataChannelIfEmpty(ReliableDataChannel, &webrtc.DataChannelInit{
+				Ordered:    &ordered,
+				Negotiated: &negotiated,
+				ID:         &id,
+			})
+		}
+		if err != nil {
+			t.params.Logger.Errorw("create migrated data channel failed", err, "label", ci.Label)
+		} else if dcExisting {
+			t.params.Logger.Debugw("existing data channel during migration", "label", dcLabel, "id", dcID)
+		} else {
+			t.params.Logger.Debugw("create migrated data channel", "label", dcLabel, "id", dcID)
+		}
+	}
+}
+
+/*
+func (p *ParticipantImpl) incActiveCounter() {
```

Comment:
can remove these

Reply:
Yep, sorry missed those. Thank you for catching. Will remove.